### PR TITLE
DAOS-1966 control: specify start rev and model name in nvme fwupdate

### PR DESCRIPTION
**Pull request by tanabarr**
Firmware updates for NVMe SSDs performed by daos_server through SPDK
specify start revision so that during distributed updates
across a number of hosts, only desired firmware update paths are
executed to avoid undesired consequences due to unsupported paths.

Model name should also be supplied so that only relevant SSDs are
updated with the provided firmware.

Signed-off-by: Tom Nabarro <tom.nabarro@intel.com>

### CODE REVIEW

File: src/control/server/storage_nvme.go [hunk 1]
```diff
@@ -345,51 +330,136 @@ func (n *nvmeStorage) Format(i int, resp *pb.FormatStorageResp) {
...
+							msgBdevFwrevStartMismatch+
+							" want %s, have %s",
+						req.Startrev, ctrlr.Fwrev))
+				continue
+			}
+
+			log.Debugf(
+				"updating firmware (%s->%s) on nvme "+
+					"controller at %s, may take several minutes!",
+				req.Startrev, ctrlr.Fwrev, pciAddr)
```
Comment (jbrosenz) [inline src/control/server/storage_nvme.go]: The message implies that we will see a change from firmware version A to B, but req.Startrev and ctrlr.Fwrev are the same value here.  Do you have access to the expected firmware value after the upgrade?  If so, then let's print that instead, or change the message so it doesn't reference the firmware versions.

Comment (tanabarr) [reply src/control/server/storage_nvme.go]: good catch

File: src/control/dmg/storage.go [hunk 2]
```diff
@@ -97,35 +99,51 @@ func (s *FormatStorCmd) Execute(args []string) error {
...
-//		Ctrlr: nil, Path: strings.TrimSpace(path), Slot: slot}, nil
-//}
+// run NVMe and SCM storage update on all connected servers
+func updateStor(params *pb.UpdateStorageParams) {
+	fmt.Println(
+		"This could be a destructive operation and storage devices " +
+			"specified in the server config file will have firmware " +
+			"updated. Please check this is a supported upgrade path " +
+			"and be patient as it may take several minutes.\n" +
+			"Are you sure you want to continue? (yes/no)")
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/dmg/storage.go]: fixed

File: src/control/dmg/shell.go [hunk 3]
```diff
@@ -24,17 +24,63 @@
...
+
+	var slot int32
+
+	if slotRaw != "" {
+		slot, err := strconv.Atoi(slotRaw)
+		if err != nil {
+			return nil, fmt.Errorf("%s is not a number", slotRaw)
+		}
+
+		if slot >= 0 && slot < 7 {
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/dmg/shell.go]: good catch I was manually testing with subcommand so missed this

File: src/control/server/mgmt_storage_test.go [hunk 4]
```diff
@@ -420,6 +420,159 @@ func TestFormatStorage(t *testing.T) {
...
+		mock := &mockUpdateStorageServer{}
+		mockWg := new(sync.WaitGroup)
+		mockWg.Add(1)
+
+		params := &pb.UpdateStorageParams{
+			Nvme: tt.nvmeParams,
+			Scm:  tt.scmParams,
+		}
+
+		go func() {
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/server/mgmt_storage_test.go]: another good catch, I've just copied and pasted from format tests where the wait group was needed

File: src/control/server/storage_nvme.go [hunk 5]
```diff
@@ -345,51 +330,136 @@ func (n *nvmeStorage) Format(i int, resp *pb.FormatStorageResp) {
...
+			if ctrlr == nil {
+				addCretUpdate(
+					pb.ResponseStatus_CTRL_ERR_NVME,
+					pciAddr+": "+msgBdevNotFound+
+						" (after update)")
+				continue
+			}
+
+			// verify controller is reporting an updated rev
+			if ctrlr.Fwrev == req.Startrev {
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/server/storage_nvme.go]: not easily without loading the firmware image and reading some header/metadata from it. will add check to make sure it's not reporting an empty string but I think printing the new rev to debug additionally should be sufficient for the moment. can definitely think about improvements for the next iteration.

**Comment by tanabarr** [comment]
> Can you update the unit tests to catch the problem this fixes? I may have missed it when looking at the existing unit test updates you made. I didn't quite see what I was looking for, so if you could point it out if you already made them, that is helpful.

this isn't a fix, this is the fwupdate feature PR, the regression fix is https://github.com/daos-stack/daos/pull/522

**Comment by johannlombardi** [comment]
> > Can you update the unit tests to catch the problem this fixes? I may have missed it when looking at the existing unit test updates you made. I didn't quite see what I was looking for, so if you could point it out if you already made them, that is helpful.
> 
> this isn't a fix, this is the fwupdate feature PR, the regression fix is #522

But, to Joel's point, PR #522 has not unit test, right?

**Comment by tanabarr** [comment]
> > > Can you update the unit tests to catch the problem this fixes? I may have missed it when looking at the existing unit test updates you made. I didn't quite see what I was looking for, so if you could point it out if you already made them, that is helpful.
> > 
> > 
> > this isn't a fix, this is the fwupdate feature PR, the regression fix is #522
> 
> But, to Joel's point, PR #522 has not unit test, right?

so to give context; we have common functions in the client package (client API) e.g. UpdateStorage() which get unit tested. Those functions are called both from subcommand Execute routines (which activate when a particular subcommand is called on cli) and from shell commands (executed from interactive shell which is launched when no sub commands are provided on cli). 

The logic surroundinyg the calls to the API in each of those cases is intentionally thin and in the case of subcommand usage there is very little to test because the parameters are basically passed directly from the third-party package routines to the client API (go-flags is well used and tested and we don't need to repeat that).

In the case of the shell, there is some logic required in prompting users to supply parameters which is as yet untested but IMO that is better covered by external framework driven tests which will supply combinations of parameters (probably something like python  pexpect). ishell package is less well maintained and tested. 

it's difficult/impossible to "unit-test" that invocation of the same command through shell or subcommand yield the same client API call as this requires an external test runner and therefore is a candidate for functional testing in CI.

The changes made in #522 simply apply the same set up before both shell and subcommand calls to client API. The only untested parts of this common appSetup routine are the configurations file processing/cli parameter overrides introduced by https://github.com/daos-stack/daos/commit/02f8b5b1799d00357d89681ce41d9f71a52293a2.

**Comment by johannlombardi** [comment]
ok. unfortunately, i cannot land due to the CI failure.

**Pull request merged**